File: rust/one_to_one_for_many.rs
use std::time::SystemTime;
use std::{thread, time};
use std::sync::{ Arc, Mutex};

use liner_broker::Liner;

fn  main() {

    let mut client1 = Liner::new("client1", "topic_client", "localhost:2255", "redis://localhost/");
    let mut client2 = Liner::new("client2", "topic_client", "localhost:2256", "redis://localhost/");
    let mut client3 = Liner::new("client3", "topic_client", "localhost:2257", "redis://localhost/");
    let mut server1 = Liner::new("server1", "topic_server1", "localhost:2258", "redis://localhost/");
    
    server1.clear_stored_messages();
    client1.clear_stored_messages();
    client2.clear_stored_messages();
    client3.clear_stored_messages();

    server1.clear_addresses_of_topic();
    client1.clear_addresses_of_topic();
    client2.clear_addresses_of_topic();
    client3.clear_addresses_of_topic();


    const MESS_SEND_COUNT: usize = 10;
    const MESS_SIZE: usize = 100;
    const SEND_CYCLE_COUNT: usize = 10;

    let receive_count_1 = Arc::new(Mutex::new(0));
    let receive_count_2 = Arc::new(Mutex::new(0));
    let receive_count_3 = Arc::new(Mutex::new(0));

    let _receive_count_1 = receive_count_1.clone();
    let _receive_count_2 = receive_count_2.clone();
    let _receive_count_3 = receive_count_3.clone();
    
    client1.run(Box::new(move |_to: &str, _from: &str,  _data: &[u8]|{
        *_receive_count_1.lock().unwrap() += 1;
    }));
    client2.run(Box::new(move |_to: &str, _from: &str,  _data: &[u8]|{
        *_receive_count_2.lock().unwrap() += 1;
    }));
    client3.run(Box::new(move |_to: &str, _from: &str,  _data: &[u8]|{
        *_receive_count_3.lock().unwrap() += 1;
    }));
    server1.run(Box::new(move |_to: &str, _from: &str,  _data: &[u8]|{}));
    
    let array = [0; MESS_SIZE];
    for _ in 0..SEND_CYCLE_COUNT{
        let send_begin = current_time_ms();
        for _ in 0..MESS_SEND_COUNT{
            server1.send_to("topic_client", array.as_slice());
        }
        let send_end = current_time_ms();
        println!("send_to {} ms", send_end - send_begin);       
    
        thread::sleep(time::Duration::from_millis(1000));
    }
    println!("receive_count_1 {}", receive_count_1.lock().unwrap());       
    println!("receive_count_2 {}", receive_count_2.lock().unwrap());       
    println!("receive_count_3 {}", receive_count_3.lock().unwrap());      
}

fn current_time_ms()->u64{ 
    SystemTime::now()
    .duration_since(SystemTime::UNIX_EPOCH)
    .unwrap()
    .as_millis() as u64
}

File: cpp/one_to_one_for_many.cpp

#include "../include/liner.h"

#include <ctime>
#include <iostream>
#include <chrono>
#include <thread>


const int MESS_SEND_COUNT = 10;
const int MESS_SIZE = 100;
const int SEND_CYCLE_COUNT = 10;

int receive_count_1 = 0, receive_count_2 = 0, receive_count_3 = 0;
clock_t send_begin = clock();
clock_t send_end = clock();

void cb_server(char* to, char* from,  char* data, size_t data_size, void* udata){
 //  std::cout << "cb1" << to << from << data << data_size << std::endl;
}

void cb_client1(char* to, char* from,  char* data, size_t data_size, void* udata){
     
    ++receive_count_1;
    std::cout << "client1 " << to << " receive_count_1 "  << receive_count_1 << std::endl;
}

void cb_client2(char* to, char* from,  char* data, size_t data_size, void* udata){
     
    ++receive_count_2;
    std::cout << "client2 " << to << " receive_count_2 "  << receive_count_2 << std::endl;
}

void cb_client3(char* to, char* from,  char* data, size_t data_size, void* udata){
     
    ++receive_count_3;
    std::cout << "client3 " << to << " receive_count_3 "  << receive_count_3 << std::endl;
}

int main(int argc, char* argv[])
{  
    auto hclient1 = ln_new_client("client1", "topic_client", "localhost:2255", "redis://localhost/");
    auto hclient2 = ln_new_client("client2", "topic_client", "localhost:2256", "redis://localhost/");
    auto hclient3 = ln_new_client("client3", "topic_client", "localhost:2257", "redis://localhost/");
    auto hserver1 = ln_new_client("server1", "topic_server", "localhost:2258", "redis://localhost/");
 
    ln_run(hclient1, cb_client1, nullptr);
    ln_run(hclient2, cb_client2, nullptr);
    ln_run(hclient3, cb_client3, nullptr);

    ln_run(hserver1, cb_server, nullptr);
 
    char data[MESS_SIZE];
    for (int i = 0; i < SEND_CYCLE_COUNT; ++i){
        send_begin = clock();
        for (int j = 0; j < MESS_SEND_COUNT; ++j){
            ln_send_to(hserver1, "topic_client", data, sizeof(data), TRUE);
        }
        send_end = clock();
        std::cout << "send_to " << 1000.0 * (send_end - send_begin) / CLOCKS_PER_SEC << " ms" << std::endl;
        std::this_thread::sleep_for(std::chrono::milliseconds(1000));
    }
    ln_delete_client(hclient1);
    ln_delete_client(hclient2);
    ln_delete_client(hclient3);
    
    ln_delete_client(hserver1);

    std::cout << "client1 receive_count_1 " << receive_count_1 << std::endl;
    std::cout << "client2 receive_count_2 " << receive_count_2 << std::endl;
    std::cout << "client3 receive_count_3 " << receive_count_3 << std::endl;
}